CMR-4247 Add support for cascading deletes to ingest

oubiwann opened the pull request:
This is a WIP; not ready for review yet.

oubiwann commented on metadata-db-app/int_test/cmr/metadata_db/int_test/concepts/delete_test.clj:
```diff
@@ -32,23 +32,40 @@
-    (testing "tag association is tombstoned after tag is deleted"
-      (is (= 200 tombstone-status))
-      (is (:deleted tombstone)))))
+(deftest tag-delete-cascades-associations
```
Note the changes here we mostly just adding `(testing ...)` calls.

oubiwann commented on metadata-db-app/src/cmr/metadata_db/data/memory_db.clj:
```diff
@@ -38,6 +39,22 @@
               (ingest-events/concept-delete-event tombstone)))
           (concat concepts tombstones))
 
+        ;; CMR-2520 Remove this case when asynchronous cascaded deletes are implemented.
```
This was adapted from analogous tag/tag-association code.

ygliuvt replied:
This code can be combined with the tag tombstone concept code above once `get-associations-for-tombstone` mutimethod is created as a result of refactoring in my other comment.

oubiwann replied:
Done.

oubiwann commented on metadata-db-app/src/cmr/metadata_db/data/oracle/concepts/variable.clj:
```diff
@@ -23,3 +25,47 @@
         [cols values] (c/concept->common-insert-args concept)]
     [(concat cols ["provider_id" "user_id" "variable_name" "measurement"])
      (concat values [provider-id user-id variable-name measurement])]))
+
```
This was adapted from analogous tag/tag-association code.

ygliuvt commented on metadata-db-app/src/cmr/metadata_db/data/oracle/concepts/variable.clj:
```diff
@@ -23,3 +25,47 @@
+                                  ;; once we move to using variable concept id.
+                                  :variable-name (:native-id variable-tombstone)}))
+
+(defn cascade-delete-variable-associations
```
How about refactoring function and `cascade-delete-tag-associations` function into a single `cascade-delete-associations` and using the concept-type on the tombstone to determine which `get-associations-for-tombstone` (new multimethod replacing `get-associations-for-variable-tombstone`) to call?

oubiwann replied:
+1

oubiwann replied:
Done.

ygliuvt commented on metadata-db-app/src/cmr/metadata_db/data/memory_db.clj:
```diff
@@ -10,35 +10,64 @@
-(defn after-save
-  "Handler for save calls. It will be passed the list of concepts and the concept that was just
-  saved. It should manipulate anything required and return the new list of concepts."
+(defmulti after-save
```
This is what I had in mind on how to refactor the `after-save` function:
{code}
(defn after-save
  "Handler for save calls. It will be passed the list of concepts and the concept that was just
  saved. It should manipulate anything required and return the new list of concepts."
  [db concepts concept]
  (cond (and (= :collection (:concept-type concept))
             (:deleted concept))
    (filter #(not= (:concept-id concept) (get-in % [:extra-fields :parent-collection-id]))
            concepts)
    
    ;; CMR-2520 Remove this case when asynchronous cascaded deletes are implemented.
    (and (or (= :tag (:concept-type concept))
             (= :variable (:concept-type concept)))
         (:deleted concept))
    (let [associations (tag/get-associations-for-tombstone db concept)
          tombstones (map (fn [ta] (-> ta
                                       (assoc :metadata "" :deleted true)
                                       (update :revision-id inc)))
                          associations)]
      ;; publish tag-association delete events
      (doseq [tombstone tombstones]
        (ingest-events/publish-event
         (:context db)
         (ingest-events/concept-delete-event tombstone)))
      (concat concepts tombstones))
    
    :else concepts))
{code}
Where `tag/get-associations-for-tombstone` is the multimethod that replaces `get-tag-associations-for-tag-tombstone` and the one for Variable.

oubiwann commented:
I think the diff for the suggested refactor is actually larger that the original change ;-) In order to accomplish the task for both external and in-memory db (which have a little bit different logic/workflows), I did the following:
* Converted the in-memory `after-save` to a multimethod
* Moved variable and tag `cascade-delete-*-associations` from their respective namespaces into `cmr.metadata-db.data.oracle.concepts` (at the end of the file)
* Split the logic trapped in that function's `let` block into a handful of utility functions, some of which are now usable by the in-memory db
* Added a new `cascade-delete-associations` at the bottom of that `concepts` ns
* Added an implementation of `c/cascade-delete-associations` in both `cmr.metadata-db.data.oracle.concepts.variable` and `cmr.metadata-db.data.oracle.concepts.tag`
* Added a utility function for delete associations in `cmr.metadata-db.data.memory-db`
* Updated the implementation of `c/after-save` in `cmr.metadata-db.data.oracle.concepts.variable`, `cmr.metadata-db.data.oracle.concepts.tag` and `cmr.metadata-db.data.memory-db`

ygliuvt commented:
I should have paired with you on the refactoring. The refactoring is meant to be minimal. :(

ygliuvt commented on metadata-db-app/src/cmr/metadata_db/data/oracle/concepts.clj:
```diff
@@ -589,3 +590,79 @@
+         (update-tombstone-associations)
+         (publish-delete-associations db provider-id))))
+
+(defmulti cascade-delete-associations
```
Really not intended to have you do this level of refactoring, but thanks.

oubiwann replied:
No worries! I shared with you on a call, but for others that read this PR -- my comment above "I think the diff for the suggested refactor is actually larger that the original change ;-)" was truly meant in good humour and not snarky at all :-)

In fact, before that refactoring I had planned on adding chunks of this code to the CMR dev pain points on the wiki page ... and now the issues I identified are no longer present -- thanks to @ygliuvt! :-D

oubiwann commented:
Thanks, all!
